# App MarketPlace Analysis 

*This project is inteded to analyze data on apps taken from the App Store and Google Play Markets to determine the characteristics of a profitable app using available data sets (10,000 Android apps and 7,000 iOS apps). This project is basic and only makes use of python built in functions and no packages such as numpy, pandas, matplotlib or scikit. However it is a good example of a data cleaning process. These packages and others will be explored in depth in other projects. 

first, I will import the datasets into lists using the csv reader tool

In [1]:
from csv import reader

opened_ios_file = open('/Users/alexkokkoros/desktop/projects/profitableapps/data/AppleStore.csv')
r1 = reader(opened_ios_file)
ios_data = list(r1)
ios_header = ios_data[0]
ios_data = ios_data[1:]

opened_android_file = open('/Users/alexkokkoros/desktop/projects/profitableapps/data/googleplaystore.csv')
r2 = reader(opened_android_file)
android_data = list(r2)
android_header = android_data[0]
android_data = android_data[1:]



next, I define a function used to explore the data sets and their dimensions

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

here, I explore the ios_data

In [3]:
explore_data(ios_data,0,5,True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16


We have 7197 iOS apps in this data set, and the columns that seem interesting are: 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre'. Not all column names are self-explanatory in this case, but details about each column can be found in the data set [documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home).

Next, I explore the android_data

In [4]:
explore_data(android_data,0,5,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 10841
Number of columns: 13


We see that the Google Play data set has 10841 apps and 13 columns. At a quick glance, the columns that might be useful for the purpose of our analysis are 'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'.

## Data Cleaning

*In this section, I will preform cleaning of the data by eliminating any missing or incorrect data.

This potential missing value in data raised in [discussion forum post](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) will be explored.

In [5]:
print(android_data[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


There is a missing entry in 'android_data[10473]' that is causing a column shift for the following columns. This row is removed. 

In [6]:
del(android_data[10472])

It was also pointed out in the discussion forum that android_data has duplicate rows, as can be seen below. There are four apps in this data set named 'Instagram'.

In [7]:
for app in android_data:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


I will now write a loop to count the number of duplicate entries in the android_data set

In [8]:
duplicate_apps = []
unique_apps = []

for app in android_data:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print("The count of unique apps in this data set: "+str(len(unique_apps)))
print("The count of duplicate apps in this data set: "+str(len(duplicate_apps)))
print("Duplicate apps names, sample: "+str(duplicate_apps[:10]))

The count of unique apps in this data set: 9659
The count of duplicate apps in this data set: 1181
Duplicate apps names, sample: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


Next, these duplicate rows will be removed. The fourth column of this data set lists the amount of reviews and app has. The instance of an app with the most reviews will be kept and its duplicates deleted. The app with the most reviews is the most recent and most accurate to use for analysis. 

A dictionary will be created to store apps and their corresponding largest value of reviews. The length of the dictionary should be equal to the count of unique values found in the previous step: 9659

In [9]:
reviews_max = {}
for app in android_data:
    name = app[0]
    n_reviews = float(app[3])
    if (name in reviews_max) and (reviews_max[name]<n_reviews):
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))

9659


The dictionary reviews_max will be used to remove the duplicate entries with lesser amount of reviews. The android_clean data set should be left with the 9659 unique apps.

In [10]:
android_clean = []
already_added = []

for app in android_data:
    name = app[0]
    n_reviews = float(app[3])
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)
    
print(len(android_clean))

9659


The 'ios_data' set fortunately has no duplicate apps. There are however, apps in both data sets directed towards non-english speakers which will not be used in our analysis. Some examples of these can be seen below. 

In [11]:
print(ios_data[813][1])
print(ios_data[6732][1])
print('\n')
print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
クリスマスミッケ/脱出ゲーム感覚の絵探しパズルゲーム


中国語 AQリスニング
لعبة تقدر تربح DZ


These non-english apps will be detected by checking the ASCII values of the characters in their name. Apps with names which contain ASCII values not within the english alphabet (0-127) will be identified and removed using the function below.

The function will return 'False' if there is more than 3 characters in the string that does not belong to the set of common English characters, otherwise it will return 'True'.

I am using 3 or more characters outside of the english language to avoid deleting apps such as 'Instachat 😜' and 'Docs To Go™ Free Office Suite' which contain one non-english character, but are still directed at english speakers. This logic check will identify apps such as '爱奇艺PPS -《欢乐颂2》电视剧热播' which are not directed at english speakers and contain 3 or more characters outside of the english language. 

In [12]:
def englishCheck(name):
    counter = 0
    for i in name:
        if ord(i)>127:
            counter += 1
            if counter > 3:
                return False
    return True

This functions is then used to create new lists with english data only. 

In [13]:
android_clean_english_apps = []
ios_english_apps = []

for app in android_clean:
    if englishCheck(app[0])==True:
        android_clean_english_apps.append(app)
        
for app in ios_data:
    if englishCheck(app[1])==True:
        ios_english_apps.append(app)
        
        
print(ios_english_apps[:5])
print(len(ios_english_apps))
print('\n')

print(android_clean_english_apps[:5])
print(len(android_clean_english_apps))

[['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'], ['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1'], ['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1'], ['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']]
6183


[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', 

We are left here with 6183 apps from the App Store and 9614 from the Google Play Store. All entries in each data set is an english app and contains no duplicates. 

For purposes of this analysis, we will only be investigating apps which are free to download. Below, I will create final lists with only free apps. 

In [14]:
android_final = []
ios_final = []

for app in android_clean_english_apps:
    if app[7] == '0':
        android_final.append(app)
        
for app in ios_english_apps:
    if app[4] == '0.0':
        ios_final.append(app)

print(len(android_final))
print(len(ios_final))

8864
3222


We are left with 8862 android apps and 3222 ios apps, all of which are english apps that are free and these lists contain no duplicates. 

## Analysis

Now that we have clean data sets, we will begin further analysis into these apps.

### Frequency Distributions of Apps by Category

Our first step of analysis will be to get an idea of what categories are most popular on each marketplace by building a frequency table of different categories.

We will do this by inspecting the 'prime_genre' column of the App Store data and the columns 'Genres' and 'Category' of the Google Play data set.

Below, I will build two functions used to create a frequency table that show percentages for certain columns of a data set and the other to display the percentages in descending order.

In [15]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Below, I will display the frequency table for the 'prime-genre' column of the ios data

In [16]:
display_table(ios_final,-5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


It is apparent that a large percentage of the apps in the App Store, more than half (58.16%) are games. No other category comes close. Entertainment apps make up the next largest portion of apps on the App Store at 7.88%, followed by Photo & Video apps at 4.96%. 

From this data, we can take away that selection of apps in the app store is dominated by apps intended for fun and entertainment (Games, Entertainment, Photo & Video) and that apps intended for specific purposes such as Navigation or Weather are rare. It should be pointed out that this data does not inform us about the popularity or amount of users a certain app has, just what types of apps are most popular in the entire library. 

Next, we will take a look at the 'Category' column of the Google Play Store

In [17]:
display_table(android_final, 1) #Category Column

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

It appears that there is a difference in the distribution of types of apps in the Google Play Store. Apps of the 'Family' and 'Game' category make up a significant portion of the distribution, with a combined 28.63%. Upon further investigation, by looking at the family section of the Google Play Store, it appears that most of these apps are games meant for young children. Even with that in mind, it seems that a larger percentage of apps are designed for practical purposes than in the App Store. This is confirmed by examing the 'Genres' column below.

In [18]:
display_table(android_final,-4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The 'Genres' category is more granular than the 'Category' column, but the exact difference is unclear. The 'Category' will be used in further analysis. 

We observed that the App Store has a larger percentage of it's distribution made up of games designed for fun and entertainment, while the Google Play Store has a more even distribution of app types. 

The next part of my analysis will investigate which apps have the most users. 

### Popular Apps by Genre on the App Store

The Google Play Store data set has a column 'Installs' which would be great to use to analyze the most popular app types. However, the App Store data set does not contain a similar column. Instead, I will use the the total number of user ratings which also indicates the amount of users an app has. In the App Store data set, this information is stored in the 'rating_count_tot' column. 

First, I will calculate the average number of user ratings per app type in the App Store data.

In [19]:
genres_ios = freq_table(ios_final, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)


Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


From these results, we can see that on average, Navigation, Social Networking, and Reference apps have the highest number of user reviews. These figures are skewed by very popular apps with a disproportionate amount of user reviews for the genre, such as Waze or Google Maps in Navigation, or Facebook, Pinterest or Skype in Social Networking. A more accurate idea of average user reviews per category in general could be obtianed by removing these wildly popular apps. This level will not be explored in this project. 

### Popular Apps by Genre in Google Play Store

The Google Play Store data set does contain information on installs, so this data will be used for the Google Play Store app popularity analysis.

The values in this data are not precise which is ok to use for purposes of this analysis. A problem is also presented with the values being stored as strings, they will have to be converted to a nuemric data type, 'float'. 

In [20]:
categories_android = freq_table(android_final, 1)


for category in categories_android:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

We can observe from this result that the communication category of apps have the most installs at 38,456,119. But, like the result from the App Store, these numbers are heavily skewed by the very popular apps, in this column. To get a more general idea of the overall popularity of apps for a certain type, these outliers would have to be removed and the analysis performed again without these values to skew the results. 

This next level of analysis will not be performed here, but will be explored further in future projects where it can be performed with more efficient tools such as pandas and scikit. 

